src/app/archive/archive_lib/processor.ml

```diff
@@ -1,264 +1,539 @@
-    let user_commands_with_times = Map.to_alist added in
-    let user_commands_with_hashes_and_times =
-      List.map user_commands_with_times ~f:(fun (user_command, times) ->
-          ( With_hash.of_data user_command
-              ~hash_data:Transaction_hash.hash_user_command
-          , times ) )
-    in
-    let user_command_hashes =
-      List.map user_commands_with_hashes_and_times
-        ~f:(fun ({With_hash.hash; _}, _) -> hash)
-    in
-    let%bind queried_existing_user_commands =
-      let graphql =
-        Graphql_query.User_commands.Query_first_seen.make
-          ~hashes:
-            ( List.map user_command_hashes ~f:Transaction_hash.to_base58_check
-            |> Array.of_list )
-          ()
-      in
-      let%map obj = Client.query graphql hasura_endpoint in
-      let list =
-        Array.map obj#user_commands ~f:(fun obj -> (obj#hash, obj#first_seen))
-        |> Array.to_list
-      in
-      Transaction_hash.Map.of_alist_exn list
+    let hash = Frozen_ledger_hash.to_string t in
+    match%bind
+      Conn.find_opt
+        (Caqti_request.find_opt Caqti_type.string Caqti_type.int
+           "SELECT id FROM snarked_ledger_hashes WHERE value = ?")
+        hash
+    with
+    | Some id ->
+        return id
+    | None ->
+        Conn.find
+          (Caqti_request.find Caqti_type.string Caqti_type.int
+             "INSERT INTO snarked_ledger_hashes (value) VALUES (?) RETURNING id")
+          hash
+end
+
+module User_command = struct
+  type t =
+    { typ: string
+    ; sender_id: int
+    ; receiver_id: int
+    ; nonce: int
+    ; amount: int option
+    ; fee: int
+    ; memo: string
+    ; hash: string }
+
+  let typ =
+    let encode t =
+      Ok
+        ( t.typ
+        , ( t.sender_id
+          , ( t.receiver_id
+            , (t.nonce, (t.amount, (t.fee, (t.memo, (t.hash, ()))))) ) ) )
     in
-    let user_commands =
-      List.map user_commands_with_hashes_and_times
-        ~f:(fun ((With_hash.{data= _; hash} as user_command_with_hash), time)
-           ->
-          let current_first_time_opt =
-            Option.join (Map.find queried_existing_user_commands hash)
-          in
-          Types.User_command.encode user_command_with_hash
-          @@ Some
-               (Option.value_map ~default:time current_first_time_opt
-                  ~f:(fun current_time -> Block_time.min current_time time)) )
+    let decode
+        ( typ
+        , ( sender_id
+          , (receiver_id, (nonce, (amount, (fee, (memo, (hash, ())))))) ) ) =
+      Ok {typ; sender_id; receiver_id; nonce; amount; fee; memo; hash}
     in
-    let graphql =
-      Graphql_query.User_commands.Insert.make
-        ~user_commands:(Array.of_list user_commands)
-        ()
+    let rep =
+      Caqti_type.(
+        tup2 string
+          (tup2 int
+             (tup2 int
+                (tup2 int
+                   (tup2 (option int)
+                      (tup2 int (tup2 string (tup2 string unit))))))))
```

Comment:
Might be worth the time to add a way to describe caqti types as hlists. This is very similar to what we do in Snarky for defining the `typ` which maps between OCaml values are Snarky vars.